sys/kern/src/arch/arm_m.rs

```diff
@@ -360,84 +459,102 @@ pub fn apply_memory_protection(task: &task::Task) {
-        // else but this routine. Here we must convert between the two -- and
-        // quickly, because this is called on every context switch.
-        //
-        // The image-generation tools check at build time that region sizes are
-        // powers of two. So, we can assume that the size has a single 1 bit. We
-        // can cheaply compute log2 of this by counting trailing zeroes, but
-        // ARMv7-M doesn't have a native instruction for that -- only leading
-        // zeroes. The equivalent using leading zeroes is
+        let data = region.arch_data;
+        // With the MPU off, there are no particular constraints on the order in
+        // which we write these fields.
         //
-        //   log2(N) = bits_in_word - 1 - clz(N)
-        //
-        // Because we want log2 _minus one_ we compute it as...
-        //
-        //   log2_m1(N) = bits_in_word - 2 - clz(N)
-        //
-        // If the size is zero or one, this subtraction will underflow. This
-        // should not occur in a valid image, but could occur due to runtime
-        // flash corruption. Any region size under 32 bytes is illegal on
-        // ARMv7-M anyway, so panicking is better than triggering possibly
-        // undefined hardware behavior.
-        //
-        // On ARMv6-M, there is no CLZ instruction either. This winds up
-        // generating decent intrinsic code for `leading_zeros` so we'll live
-        // with it.
-        let l2size = 30 - region.size.leading_zeros();
-
-        // Region attribute and size register; note that enable (bit 0) is not
-        // set here, because it's possible to hard-fault midway through region
-        // configuration if address and size are incompatible while the region
-        // is enabled.
-        let rasr =
-            (xn as u32) << 28 | ap << 24 | tex << 19 | scb << 16 | l2size << 1;
+        // Safety: we're messing with memory protection, so from the API's point
+        // of view this is very unsafe. But we're loading values generated by
+        // our (trusted) build script, which only affect tasks and not us. So
+        // this should be safe by default.
         unsafe {
-            mpu.rnr.write(i as u32); // Select the region
-            mpu.rasr.write(rasr); // configure, but leave disabled
-            mpu.rbar.write(region.base); // set region address
-            mpu.rasr.write(rasr | 1); // enable the region
+            // Select a region.
+            mpu.rnr.write(i as u32);
+            // Set region base address.
+            mpu.rbar.write(data.rbar);
+            // Configure the region.
+            mpu.rasr.write(data.rasr);
         }
     }
+
+    // Turn MPU back on.
+    //
+    // Safety: same as above, has no safety implications really.
+    unsafe {
+        mpu.ctrl.write(0b101);
+    }
+}
+
+#[cfg(armv8m)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
+pub struct RegionDescExt {
+    rlar_disabled: u32,
```

Comment:
Can you add add comment explaining this is the RLAR with everything except the `enable` bit set?  It took me a bit to figure out that was what `rlar_disabled` meant

Reply:
I just had to go convince myself that we _need_ `rlar_disabled` so I could explain it ... I think we actually don't, but I'll leave it in in the interest of not breaking things. I will add an explanatory comment and a TODO.